File: js/sms/status-callback.js
const http = require('http');
const express = require('express');

const PORT = 8767;
const app = express();
app.use(express.urlencoded({ extended: false }));

app.post('/sms', (request, response) => {
	console.log(
		`Incoming message from ${request.body.From}: ${request.body.Body}`
	);
	response.type('text/xml');
	// adding the Action attribute and pointing it to relative /status link
	// will do /status route under the hood when the message is sent. 
	response.send(`
    <Response>
		<Message action="/status">
			This is my reply
		</Message>
	</Response>
	`);
});

app.post('/status', request => {
	console.log('Status update received');
	console.log('Message status: ', request.body.MessageStatus);
	console.log('Message SID: ', request.body.MessageSid);
});

const server = http.createServer(app);
server.listen(PORT, () =>
	console.log(`Express server listening on localhost:${PORT}`)
);

File: py/sms/status-callback.py
from flask import Flask, request
from twilio.twiml.messaging_response import MessagingResponse
import logging

# The stautscallback in pytohin DOESNT require logging, but i saw a tutorial that sues it so why not right?
logging.basicConfig(level=logging.INFO)

app = Flask(__name__)


@app.route("/sms", methods=['GET', 'POST'])
def sms_ahoy_reply():
    print(
        f'Incoming message from {request.values.get("From")}: ${request.values.get("Body")}')
    resp = MessagingResponse()
    resp.redirect("/status")
    return str(resp)

# we just need post here.


@app.route("/status", methods=['POST'])
def status_ahoy_reply():
    message_sid = request.values.get('MessageSid', None)
    message_status = request.values.get('MessageStatus', None)
    logging.info('SID: {}, Status: {}'.format(message_sid, message_status))

# The DEF has to reutnr SOMETHING. so we are returning a stauts of 204, while logging the things in logging.
    return ('', 204)


if __name__ == "__main__":
    app.run(port=8767)
